uacp/endpoint_test.go

```diff
@@ -5,72 +5,77 @@
 package uacp
 
 import (
-	"net"
+	"context"
+	"net/url"
 	"testing"
 )
 
 func TestResolveEndpoint(t *testing.T) {
 	cases := []struct {
 		input   string
 		network string
-		addr    *net.TCPAddr
+		u       *url.URL
 		errStr  string
 	}{
 		{ // Valid, full EndpointURL
 			"opc.tcp://10.0.0.1:4840/foo/bar",
 			"tcp",
-			&net.TCPAddr{
-				IP:   net.IP([]byte{0x0a, 0x00, 0x00, 0x01}),
-				Port: 4840,
+			&url.URL{
+				Scheme: "opc.tcp",
+				Host:   "10.0.0.1:4840",
+				Path:   "/foo/bar",
 			},
 			"",
 		},
 		{ // Valid, port number omitted
 			"opc.tcp://10.0.0.1/foo/bar",
 			"tcp",
-			&net.TCPAddr{
-				IP:   net.IP([]byte{0x0a, 0x00, 0x00, 0x01}),
-				Port: 4840,
+			&url.URL{
+				Scheme: "opc.tcp",
+				Host:   "10.0.0.1:4840",
+				Path:   "/foo/bar",
 			},
 			"",
 		},
 		{ // Valid, hostname resolved
 			"opc.tcp://localhost:4840/foo/bar",
 			"tcp",
-			&net.TCPAddr{
-				IP:   net.IP([]byte{0x7f, 0x00, 0x00, 0x01}),
-				Port: 4840,
+			&url.URL{
+				Scheme: "opc.tcp",
+				Host:   "127.0.0.1:4840",
+				Path:   "/foo/bar",
 			},
 			"",
 		},
-		{ // Invalid, schema is not "opc.tcp://"
+		{ // Invalid, scheme is not "opc.tcp://"
 			"tcp://10.0.0.1:4840/foo/bar",
 			"",
 			nil,
-			"invalid endpoint tcp://10.0.0.1:4840/foo/bar",
+			"unsupported scheme tcp",
 		},
-		{ // Invalid, bad formatted schema
-			"opc.tcp:/10.0.0.1:4840/foo/bar",
+		{ // Invalid, bad formatted endpoint
+			"opc.tcp://10.0.0.1:x4840/foo/bar",
 			"",
 			nil,
-			"could not resolve address foo:4840",
+			"parse opc.tcp://10.0.0.1:x4840/foo/bar: invalid port \":x4840\" after host",
 		},
 	}
 
 	for _, c := range cases {
 		var errStr string
-		network, addr, err := ResolveEndpoint(c.input)
+		network, u, err := ResolveEndpoint(context.Background(), c.input)
 		if err != nil {
 			errStr = err.Error()
-		}
-		if got, want := network, c.network; got != want {
-			t.Fatalf("got network %q want %q", got, want)
-		}
-		if got, want := addr.String(), c.addr.String(); got != want {
-			t.Fatalf("got addr %q want %q", got, want)
-		}
-		if got, want := errStr, c.errStr; got != want {
-			t.Fatalf("got error %q want %q", got, want)
+			if got, want := errStr, c.errStr; got != want {
```

Comment:
This handles both the `nil` and `not-nil` case without the `if` statement.

```go
if got, want := err, tt.err; !reflect.DeepEqual(got, want) {
    t.Fatalf("got error %v want %v", got, want)
}
```

Reply:
I was getting a panic above on `errStr = err.Error()` since `err` was `nil`.

Reply:
yes, you need to refactor the test to have an `error` instead of an `errStr` in the test case definition.

Reply:
we now have `errors.Equal`